lib/Artifact.ts

```diff
@@ -0,0 +1,146 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    EventFired,
+    HandlerContext,
+    HandlerResult,
+    Success,
+} from "@atomist/automation-client";
+import { subscription } from "@atomist/automation-client/lib/graph/graphQL";
+import {
+    addressChannelsFor,
+    ArtifactGoal,
+    ArtifactListenerInvocation,
+    ArtifactListenerRegisterable,
+    DefaultGoalNameGenerator,
+    findSdmGoalOnCommit,
+    FulfillableGoalDetails,
+    FulfillableGoalWithRegistrations,
+    getGoalDefinitionFrom,
+    Goal,
+    logger,
+    PushListenerInvocation,
+    SdmGoalState,
+    SideEffect,
+    toArtifactListenerRegistration,
+    updateGoal,
+} from "@atomist/sdm";
+import * as _ from "lodash";
+import { OnImageLinked } from "./typings/types";
+
+export interface ArtifactRegistration {
+    listeners?: ArtifactListenerRegisterable[];
+}
+
+export class Artifact extends FulfillableGoalWithRegistrations<ArtifactRegistration> {
+    constructor(goalDetailsOrUniqueName: FulfillableGoalDetails | string
+                    = DefaultGoalNameGenerator.generateName("artifact"),
+                ...dependsOn: Goal[]) {
+
+        super({
+            ...ArtifactGoal.definition,
+            ...getGoalDefinitionFrom(goalDetailsOrUniqueName, DefaultGoalNameGenerator.generateName("artifact")),
+            displayName: "artifact",
+        }, ...dependsOn);
+
+        const fulfillment: SideEffect = {name: "build"};
+        this.addFulfillment(fulfillment);
+        this.sdm.addEvent({
+            name: "FindOnArtifactImageLinked",
+            subscription: subscription("OnImageLinked"),
+            listener: (event, context) => this.handle(event, context, this),
+        });
+    }
+
+    private async handle(event: EventFired<OnImageLinked.Subscription>,
+                         context: HandlerContext,
+                         goal: Artifact): Promise<HandlerResult> {
+        const imageLinked = event.data.ImageLinked[0];
+        const commit = imageLinked.commit;
+        const image = imageLinked.image;
+        const id = goal.sdm.configuration.sdm.repoRefResolver.toRemoteRepoRef(
+            commit.repo,
+            {
+                sha: commit.sha,
+                branch: commit.pushes[0].branch,
+            });
+
+        const artifactSdmGoal = await findSdmGoalOnCommit(
+            context,
+            id,
+            commit.repo.org.provider.providerId,
+            this);
+        if (!artifactSdmGoal) {
+            logger.debug("Context %s not found for %j", goal.context, id);
```

Comment:
That log statement should be removed. It is not needed and also a left over from our previous use of context everywhere.